client/src/main/java/org/apache/celeborn/client/read/WorkerPartitionReader.java

```diff
@@ -74,25 +82,54 @@ public void onSuccess(int chunkIndex, ManagedBuffer buffer) {
           }
 
           @Override
-          public void onFailure(int chunkIndex, Throwable e) {
-            String errorMsg = "Fetch chunk " + chunkIndex + " failed.";
-            logger.error(errorMsg, e);
-            exception.set(new IOException(errorMsg, e));
+          public void onFailure(int chunkIndex, Throwable e, PartitionLocation location) {
+            readableLocations.remove(location);
+            if (readableLocations.isEmpty()) {
+              String errorMsg = "Fetch chunk " + chunkIndex + " failed.";
+              logger.error(errorMsg, e);
+              exception.set(new IOException(errorMsg, e));
+            } else {
+              if (!readableLocations.contains(location)) {
+                // This branch means current location is failed and changed to its peer,
+                // so we should ignore it. This logic is to ensure that peer will be retried only
+                // once.
+                return;
+              }
+              try {
+                synchronized (this) {
+                  client.set(
+                      new ChunkClient(
+                          conf,
+                          shuffleKey,
+                          location.getPeer(),
+                          this,
+                          clientFactory,
+                          startMapIndex,
+                          endMapIndex));
+                  currentChunkIndex.set(0);
+                  returnedChunks.set(0);
+                  numChunks.set(client.get().openChunks());
+                }
+              } catch (IOException e1) {
+                logger.error(e1.getMessage(), e1);
+                exception.set(new IOException(e1.getMessage(), e1));
+              }
+            }
           }
         };
-    client =
-        new RetryingChunkClient(
-            conf, shuffleKey, location, callback, clientFactory, startMapIndex, endMapIndex);
-    numChunks = client.openChunks();
+    client.set(
+        new ChunkClient(
+            conf, shuffleKey, location, callback, clientFactory, startMapIndex, endMapIndex));
+    numChunks.set(client.get().openChunks());
   }
 
-  public boolean hasNext() {
-    return returnedChunks < numChunks;
+  public synchronized boolean hasNext() {
+    return returnedChunks.get() < numChunks.get();
   }
 
-  public ByteBuf next() throws IOException {
+  public synchronized ByteBuf next() throws IOException {
     checkException();
-    if (chunkIndex < numChunks) {
+    if (currentChunkIndex.get() < numChunks.get()) {
```

Comment:
There will be concurrent issue here. If results is empty and fetch chunk fail, it will set exception since this method is sync and on failure will wait